man/lhash.3
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "lhash 3"
.TH lhash 3 "2014-07-11" "LibreSSL" "LibreSSL"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
lh_new, lh_free, lh_insert, lh_delete, lh_retrieve, lh_doall, lh_doall_arg,
lh_error \- dynamic hash table
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& #include <openssl/lhash.h>
\&
\& DECLARE_LHASH_OF(<type>);
\&
\& LHASH *lh_<type>_new();
\& void lh_<type>_free(LHASH_OF(<type> *table);
\&
\& <type> *lh_<type>_insert(LHASH_OF(<type> *table, <type> *data);
\& <type> *lh_<type>_delete(LHASH_OF(<type> *table, <type> *data);
\& <type> *lh_retrieve(LHASH_OF<type> *table, <type> *data);
\&
\& void lh_<type>_doall(LHASH_OF(<type> *table, LHASH_DOALL_FN_TYPE func);
\& void lh_<type>_doall_arg(LHASH_OF(<type> *table, LHASH_DOALL_ARG_FN_TYPE func,
\&          <type2>, <type2> *arg);
\&
\& int lh_<type>_error(LHASH_OF(<type> *table);
\&
\& typedef int (*LHASH_COMP_FN_TYPE)(const void *, const void *);
\& typedef unsigned long (*LHASH_HASH_FN_TYPE)(const void *);
\& typedef void (*LHASH_DOALL_FN_TYPE)(const void *);
\& typedef void (*LHASH_DOALL_ARG_FN_TYPE)(const void *, const void *);
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This library implements type-checked dynamic hash tables. The hash
table entries can be arbitrary structures. Usually they consist of key
and value fields.
.PP
lh_<type>\fI_new()\fR creates a new \fB\s-1LHASH_OF\s0(<type\fR> structure to store
arbitrary data entries, and provides the 'hash' and 'compare'
callbacks to be used in organising the table's entries.  The \fBhash\fR
callback takes a pointer to a table entry as its argument and returns
an unsigned long hash value for its key field.  The hash value is
normally truncated to a power of 2, so make sure that your hash
function returns well mixed low order bits.  The \fBcompare\fR callback
takes two arguments (pointers to two hash table entries), and returns
0 if their keys are equal, non-zero otherwise.  If your hash table
will contain items of some particular type and the \fBhash\fR and
\&\fBcompare\fR callbacks hash/compare these types, then the
\&\fB\s-1DECLARE_LHASH_HASH_FN\s0\fR and \fB\s-1IMPLEMENT_LHASH_COMP_FN\s0\fR macros can be
used to create callback wrappers of the prototypes required by
lh_<type>\fI_new()\fR.  These provide per-variable casts before calling the
type-specific callbacks written by the application author.  These
macros, as well as those used for the \*(L"doall\*(R" callbacks, are defined
as;
.PP
.Vb 7
\& #define DECLARE_LHASH_HASH_FN(name, o_type) \e
\&         unsigned long name##_LHASH_HASH(const void *);
\& #define IMPLEMENT_LHASH_HASH_FN(name, o_type) \e
\&         unsigned long name##_LHASH_HASH(const void *arg) { \e
\&                 const o_type *a = arg; \e
\&                 return name##_hash(a); }
\& #define LHASH_HASH_FN(name) name##_LHASH_HASH
\&
\& #define DECLARE_LHASH_COMP_FN(name, o_type) \e
\&         int name##_LHASH_COMP(const void *, const void *);
\& #define IMPLEMENT_LHASH_COMP_FN(name, o_type) \e
\&         int name##_LHASH_COMP(const void *arg1, const void *arg2) { \e
\&                 const o_type *a = arg1;                    \e
\&                 const o_type *b = arg2; \e
\&                 return name##_cmp(a,b); }
\& #define LHASH_COMP_FN(name) name##_LHASH_COMP
\&
\& #define DECLARE_LHASH_DOALL_FN(name, o_type) \e
\&         void name##_LHASH_DOALL(void *);
\& #define IMPLEMENT_LHASH_DOALL_FN(name, o_type) \e
\&         void name##_LHASH_DOALL(void *arg) { \e
\&                 o_type *a = arg; \e
\&                 name##_doall(a); }
\& #define LHASH_DOALL_FN(name) name##_LHASH_DOALL
\&
\& #define DECLARE_LHASH_DOALL_ARG_FN(name, o_type, a_type) \e
\&         void name##_LHASH_DOALL_ARG(void *, void *);
\& #define IMPLEMENT_LHASH_DOALL_ARG_FN(name, o_type, a_type) \e
\&         void name##_LHASH_DOALL_ARG(void *arg1, void *arg2) { \e
\&                 o_type *a = arg1; \e
\&                 a_type *b = arg2; \e
\&                 name##_doall_arg(a, b); }
\& #define LHASH_DOALL_ARG_FN(name) name##_LHASH_DOALL_ARG
\&
\& An example of a hash table storing (pointers to) structures of type \*(AqSTUFF\*(Aq
\& could be defined as follows;
\&
\& /* Calculates the hash value of \*(Aqtohash\*(Aq (implemented elsewhere) */
\& unsigned long STUFF_hash(const STUFF *tohash);
\& /* Orders \*(Aqarg1\*(Aq and \*(Aqarg2\*(Aq (implemented elsewhere) */
\& int stuff_cmp(const STUFF *arg1, const STUFF *arg2);
\& /* Create the type\-safe wrapper functions for use in the LHASH internals */
\& static IMPLEMENT_LHASH_HASH_FN(stuff, STUFF);
\& static IMPLEMENT_LHASH_COMP_FN(stuff, STUFF);
\& /* ... */
\& int main(int argc, char *argv[]) {
\&         /* Create the new hash table using the hash/compare wrappers */
\&         LHASH_OF(STUFF) *hashtable = lh_STUFF_new(LHASH_HASH_FN(STUFF_hash),
\&                                   LHASH_COMP_FN(STUFF_cmp));
\&         /* ... */
\& }
.Ve
.PP
lh_<type>\fI_free()\fR frees the \fB\s-1LHASH_OF\s0(<type\fR> structure
\&\fBtable\fR. Allocated hash table entries will not be freed; consider
using lh_<type>\fI_doall()\fR to deallocate any remaining entries in the
hash table (see below).
.PP
lh_<type>\fI_insert()\fR inserts the structure pointed to by \fBdata\fR into
\&\fBtable\fR.  If there already is an entry with the same key, the old
value is replaced. Note that lh_<type>\fI_insert()\fR stores pointers, the
data are not copied.
.PP
lh_<type>\fI_delete()\fR deletes an entry from \fBtable\fR.
.PP
lh_<type>\fI_retrieve()\fR looks up an entry in \fBtable\fR. Normally, \fBdata\fR
is a structure with the key field(s) set; the function will return a
pointer to a fully populated structure.
.PP
lh_<type>\fI_doall()\fR will, for every entry in the hash table, call
\&\fBfunc\fR with the data item as its parameter.  For lh_<type>\fI_doall()\fR
and lh_<type>\fI_doall_arg()\fR, function pointer casting should be avoided
in the callbacks (see \fB\s-1NOTE\s0\fR) \- instead use the declare/implement
macros to create type-checked wrappers that cast variables prior to
calling your type-specific callbacks.  An example of this is
illustrated here where the callback is used to cleanup resources for
items in the hash table prior to the hashtable itself being
deallocated:
.PP
.Vb 9
\& /* Cleans up resources belonging to \*(Aqa\*(Aq (this is implemented elsewhere) */
\& void STUFF_cleanup_doall(STUFF *a);
\& /* Implement a prototype\-compatible wrapper for "STUFF_cleanup" */
\& IMPLEMENT_LHASH_DOALL_FN(STUFF_cleanup, STUFF)
\&         /* ... then later in the code ... */
\& /* So to run "STUFF_cleanup" against all items in a hash table ... */

NEWS

  NEWS
  ====

  This file gives a brief overview of the major changes between each OpenSSL
  release. For more details please read the CHANGES file.

  Major changes between OpenSSL 1.0.2 and OpenSSL 1.0.2a [under development]

      o OpenSSL 1.0.2 ClientHello sigalgs DoS fix (CVE-2015-0291)
      o Multiblock corrupted pointer fix (CVE-2015-0290)
      o Segmentation fault in DTLSv1_listen fix (CVE-2015-0207)
      o Segmentation fault in ASN1_TYPE_cmp fix (CVE-2015-0286)
      o Segmentation fault for invalid PSS parameters fix (CVE-2015-0208)
      o ASN.1 structure reuse memory corruption fix (CVE-2015-0287)
      o PKCS7 NULL pointer dereferences fix (CVE-2015-0289)
      o DoS via reachable assert in SSLv2 servers fix (CVE-2015-0293)
      o Empty CKE with client auth and DHE fix (CVE-2015-1787)
      o Handshake with unseeded PRNG fix (CVE-2015-0285)
      o Use After Free following d2i_ECPrivatekey error fix (CVE-2015-0209)
      o X509_to_X509_REQ NULL pointer deref fix (CVE-2015-0288)
      o Removed the export ciphers from the DEFAULT ciphers

  Major changes between OpenSSL 1.0.1l and OpenSSL 1.0.2 [22 Jan 2015]:

      o Suite B support for TLS 1.2 and DTLS 1.2
      o Support for DTLS 1.2
      o TLS automatic EC curve selection.
      o API to set TLS supported signature algorithms and curves
      o SSL_CONF configuration API.
      o TLS Brainpool support.
      o ALPN support.
      o CMS support for RSA-PSS, RSA-OAEP, ECDH and X9.42 DH.

  Major changes between OpenSSL 1.0.1k and OpenSSL 1.0.1l [15 Jan 2015]

      o Build fixes for the Windows and OpenVMS platforms

  Major changes between OpenSSL 1.0.1j and OpenSSL 1.0.1k [8 Jan 2015]

      o Fix for CVE-2014-3571
      o Fix for CVE-2015-0206
      o Fix for CVE-2014-3569
      o Fix for CVE-2014-3572
      o Fix for CVE-2015-0204
      o Fix for CVE-2015-0205
      o Fix for CVE-2014-8275
      o Fix for CVE-2014-3570

  Major changes between OpenSSL 1.0.1i and OpenSSL 1.0.1j [15 Oct 2014]

      o Fix for CVE-2014-3513
      o Fix for CVE-2014-3567
      o Mitigation for CVE-2014-3566 (SSL protocol vulnerability)
      o Fix for CVE-2014-3568

  Major changes between OpenSSL 1.0.1h and OpenSSL 1.0.1i [6 Aug 2014]

      o Fix for CVE-2014-3512
      o Fix for CVE-2014-3511
      o Fix for CVE-2014-3510
      o Fix for CVE-2014-3507
      o Fix for CVE-2014-3506
      o Fix for CVE-2014-3505
      o Fix for CVE-2014-3509
      o Fix for CVE-2014-5139
      o Fix for CVE-2014-3508

  Major changes between OpenSSL 1.0.1g and OpenSSL 1.0.1h [5 Jun 2014]

      o Fix for CVE-2014-0224
      o Fix for CVE-2014-0221
      o Fix for CVE-2014-0198
      o Fix for CVE-2014-0195
      o Fix for CVE-2014-3470
      o Fix for CVE-2010-5298

  Major changes between OpenSSL 1.0.1f and OpenSSL 1.0.1g [7 Apr 2014]

      o Fix for CVE-2014-0160
      o Add TLS padding extension workaround for broken servers.
      o Fix for CVE-2014-0076

  Major changes between OpenSSL 1.0.1e and OpenSSL 1.0.1f [6 Jan 2014]

      o Don't include gmt_unix_time in TLS server and client random values
      o Fix for TLS record tampering bug CVE-2013-4353
      o Fix for TLS version checking bug CVE-2013-6449
      o Fix for DTLS retransmission bug CVE-2013-6450

  Major changes between OpenSSL 1.0.1d and OpenSSL 1.0.1e [11 Feb 2013]:

      o Corrected fix for CVE-2013-0169

  Major changes between OpenSSL 1.0.1c and OpenSSL 1.0.1d [4 Feb 2013]:

      o Fix renegotiation in TLS 1.1, 1.2 by using the correct TLS version.
      o Include the fips configuration module.
      o Fix OCSP bad key DoS attack CVE-2013-0166
      o Fix for SSL/TLS/DTLS CBC plaintext recovery attack CVE-2013-0169
      o Fix for TLS AESNI record handling flaw CVE-2012-2686

  Major changes between OpenSSL 1.0.1b and OpenSSL 1.0.1c [10 May 2012]:

      o Fix TLS/DTLS record length checking bug CVE-2012-2333
      o Don't attempt to use non-FIPS composite ciphers in FIPS mode.

  Major changes between OpenSSL 1.0.1a and OpenSSL 1.0.1b [26 Apr 2012]:

      o Fix compilation error on non-x86 platforms.
      o Make FIPS capable OpenSSL ciphers work in non-FIPS mode.
      o Fix SSL_OP_NO_TLSv1_1 clash with SSL_OP_ALL in OpenSSL 1.0.0

  Major changes between OpenSSL 1.0.1 and OpenSSL 1.0.1a [19 Apr 2012]:

      o Fix for ASN1 overflow bug CVE-2012-2110
      o Workarounds for some servers that hang on long client hellos.
      o Fix SEGV in AES code.

  Major changes between OpenSSL 1.0.0h and OpenSSL 1.0.1 [14 Mar 2012]:

      o TLS/DTLS heartbeat support.
      o SCTP support.
      o RFC 5705 TLS key material exporter.
      o RFC 5764 DTLS-SRTP negotiation.
      o Next Protocol Negotiation.
      o PSS signatures in certificates, requests and CRLs.
      o Support for password based recipient info for CMS.
      o Support TLS v1.2 and TLS v1.1.
      o Preliminary FIPS capability for unvalidated 2.0 FIPS module.
      o SRP support.

  Major changes between OpenSSL 1.0.0g and OpenSSL 1.0.0h [12 Mar 2012]:

      o Fix for CMS/PKCS#7 MMA CVE-2012-0884
      o Corrected fix for CVE-2011-4619
      o Various DTLS fixes.

  Major changes between OpenSSL 1.0.0f and OpenSSL 1.0.0g [18 Jan 2012]:

      o Fix for DTLS DoS issue CVE-2012-0050


crypto/lhash/lhash.c
/* $OpenBSD: lhash.c,v 1.16 2014/07/09 11:10:51 bcook Exp $ */
/* Copyright (C) 1995-1998 Eric Young (eay@cryptsoft.com)
 * All rights reserved.
 *
 * This package is an SSL implementation written
 * by Eric Young (eay@cryptsoft.com).
 * The implementation was written so as to conform with Netscapes SSL.
 *
 * This library is free for commercial and non-commercial use as long as
 * the following conditions are aheared to.  The following conditions
 * apply to all code found in this distribution, be it the RC4, RSA,
 * lhash, DES, etc., code; not just the SSL code.  The SSL documentation
 * included with this distribution is covered by the same copyright terms
 * except that the holder is Tim Hudson (tjh@cryptsoft.com).
 *
 * Copyright remains Eric Young's, and as such any Copyright notices in
 * the code are not to be removed.
 * If this package is used in a product, Eric Young should be given attribution
 * as the author of the parts of the library used.
 * This can be in the form of a textual message at program startup or
 * in documentation (online or textual) provided with the package.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *    "This product includes cryptographic software written by
 *     Eric Young (eay@cryptsoft.com)"
 *    The word 'cryptographic' can be left out if the rouines from the library
 *    being used are not cryptographic related :-).
 * 4. If you include any Windows specific code (or a derivative thereof) from
 *    the apps directory (application code) you must include an acknowledgement:
 *    "This product includes software written by Tim Hudson (tjh@cryptsoft.com)"
 *
 * THIS SOFTWARE IS PROVIDED BY ERIC YOUNG ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * The licence and distribution terms for any publically available version or
 * derivative of this code cannot be changed.  i.e. this code cannot simply be
 * copied and put under another distribution licence
 * [including the GNU Public Licence.]
 */

/* Code for dynamic hash table routines
 * Author - Eric Young v 2.0
 *
 * 2.2 eay - added #include "crypto.h" so the memory leak checking code is
 *	     present. eay 18-Jun-98
 *
 * 2.1 eay - Added an 'error in last operation' flag. eay 6-May-98
 *
 * 2.0 eay - Fixed a bug that occurred when using lh_delete
 *	     from inside lh_doall().  As entries were deleted,
 *	     the 'table' was 'contract()ed', making some entries
 *	     jump from the end of the table to the start, there by
 *	     skipping the lh_doall() processing. eay - 4/12/95
 *
 * 1.9 eay - Fixed a memory leak in lh_free, the LHASH_NODEs
 *	     were not being free()ed. 21/11/95
 *
 * 1.8 eay - Put the stats routines into a separate file, lh_stats.c
 *	     19/09/95
 *
 * 1.7 eay - Removed the fputs() for realloc failures - the code
 *           should silently tolerate them.  I have also fixed things
 *           lint complained about 04/05/95
 *
 * 1.6 eay - Fixed an invalid pointers in contract/expand 27/07/92
 *
 * 1.5 eay - Fixed a misuse of realloc in expand 02/03/1992
 *
 * 1.4 eay - Fixed lh_doall so the function can call lh_delete 28/05/91
 *
 * 1.3 eay - Fixed a few lint problems 19/3/1991
 *
 * 1.2 eay - Fixed lh_doall problem 13/3/1991
 *
 * 1.1 eay - Added lh_doall
 *
 * 1.0 eay - First version
 */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include <openssl/opensslconf.h>

#include <openssl/crypto.h>
#include <openssl/lhash.h>

#undef MIN_NODES
#define MIN_NODES	16
#define UP_LOAD		(2*LH_LOAD_MULT) /* load times 256  (default 2) */
#define DOWN_LOAD	(LH_LOAD_MULT)   /* load times 256  (default 1) */

static void expand(_LHASH *lh);
static void contract(_LHASH *lh);
static LHASH_NODE **getrn(_LHASH *lh, const void *data, unsigned long *rhash);

_LHASH *
lh_new(LHASH_HASH_FN_TYPE h, LHASH_COMP_FN_TYPE c)
{
	_LHASH *ret;
	int i;

	if ((ret = malloc(sizeof(_LHASH))) == NULL)
		goto err0;
	if ((ret->b = reallocarray(NULL, MIN_NODES, sizeof(LHASH_NODE *))) == NULL)
		goto err1;
	for (i = 0; i < MIN_NODES; i++)
		ret->b[i] = NULL;
	ret->comp = ((c == NULL) ? (LHASH_COMP_FN_TYPE)strcmp : c);
	ret->hash = ((h == NULL) ? (LHASH_HASH_FN_TYPE)lh_strhash : h);
	ret->num_nodes = MIN_NODES / 2;
	ret->num_alloc_nodes = MIN_NODES;
	ret->p = 0;
	ret->pmax = MIN_NODES / 2;
	ret->up_load = UP_LOAD;
	ret->down_load = DOWN_LOAD;
	ret->num_items = 0;

	ret->num_expands = 0;
	ret->num_expand_reallocs = 0;
	ret->num_contracts = 0;
	ret->num_contract_reallocs = 0;
	ret->num_hash_calls = 0;
	ret->num_comp_calls = 0;
	ret->num_insert = 0;
	ret->num_replace = 0;
	ret->num_delete = 0;
	ret->num_no_delete = 0;
	ret->num_retrieve = 0;
	ret->num_retrieve_miss = 0;
	ret->num_hash_comps = 0;

	ret->error = 0;
	return (ret);

err1:
	free(ret);
err0:
	return (NULL);
}

void
lh_free(_LHASH *lh)
{
	unsigned int i;
	LHASH_NODE *n, *nn;

	if (lh == NULL)
		return;

	for (i = 0; i < lh->num_nodes; i++) {
		n = lh->b[i];
		while (n != NULL) {
			nn = n->next;
			free(n);
			n = nn;
		}
	}
	free(lh->b);
	free(lh);
}

void *
lh_insert(_LHASH *lh, void *data)
{
	unsigned long hash;
	LHASH_NODE *nn, **rn;
	void *ret;

	lh->error = 0;
	if (lh->up_load <= (lh->num_items * LH_LOAD_MULT / lh->num_nodes))
		expand(lh);

	rn = getrn(lh, data, &hash);

	if (*rn == NULL) {
		if ((nn = malloc(sizeof(LHASH_NODE))) == NULL) {
			lh->error++;
			return (NULL);
		}
		nn->data = data;
		nn->next = NULL;
#ifndef OPENSSL_NO_HASH_COMP
		nn->hash = hash;
#endif
		*rn = nn;
		ret = NULL;
		lh->num_insert++;
		lh->num_items++;
	}
	else /* replace same key */
	{
		ret = (*rn)->data;
		(*rn)->data = data;
		lh->num_replace++;
	}
	return (ret);
}

void *
lh_delete(_LHASH *lh, const void *data)
{
	unsigned long hash;
	LHASH_NODE *nn, **rn;
	void *ret;

	lh->error = 0;
	rn = getrn(lh, data, &hash);

	if (*rn == NULL) {
		lh->num_no_delete++;
		return (NULL);
	} else {
		nn= *rn;
		*rn = nn->next;
		ret = nn->data;
		free(nn);
		lh->num_delete++;
	}

	lh->num_items--;
	if ((lh->num_nodes > MIN_NODES) &&
	    (lh->down_load >= (lh->num_items * LH_LOAD_MULT / lh->num_nodes)))
		contract(lh);

	return (ret);
}

void *
lh_retrieve(_LHASH *lh, const void *data)
{
	unsigned long hash;
	LHASH_NODE **rn;
	void *ret;

	lh->error = 0;
	rn = getrn(lh, data, &hash);

	if (*rn == NULL) {
		lh->num_retrieve_miss++;
		return (NULL);
	} else {
		ret = (*rn)->data;
		lh->num_retrieve++;
	}
	return (ret);
}

static void
doall_util_fn(_LHASH *lh, int use_arg, LHASH_DOALL_FN_TYPE func,
    LHASH_DOALL_ARG_FN_TYPE func_arg, void *arg)
{
	int i;
	LHASH_NODE *a, *n;

	if (lh == NULL)
		return;

	/* reverse the order so we search from 'top to bottom'
	 * We were having memory leaks otherwise */
	for (i = lh->num_nodes - 1; i >= 0; i--) {
		a = lh->b[i];
		while (a != NULL) {
			/* 28/05/91 - eay - n added so items can be deleted
			 * via lh_doall */
			/* 22/05/08 - ben - eh? since a is not passed,
			 * this should not be needed */
			n = a->next;
			if (use_arg)
				func_arg(a->data, arg);
			else
				func(a->data);
			a = n;
		}
	}
}

void
lh_doall(_LHASH *lh, LHASH_DOALL_FN_TYPE func)
{
	doall_util_fn(lh, 0, func, (LHASH_DOALL_ARG_FN_TYPE)0, NULL);
}

void
lh_doall_arg(_LHASH *lh, LHASH_DOALL_ARG_FN_TYPE func, void *arg)
{
	doall_util_fn(lh, 1, (LHASH_DOALL_FN_TYPE)0, func, arg);
}

static void
expand(_LHASH *lh)
{
	LHASH_NODE **n, **n1, **n2, *np;
	unsigned int p, i, j;
	unsigned long hash, nni;

	lh->num_nodes++;
	lh->num_expands++;
	p = (int)lh->p++;
	n1 = &(lh->b[p]);
	n2 = &(lh->b[p + (int)lh->pmax]);
	*n2 = NULL;        /* 27/07/92 - eay - undefined pointer bug */
	nni = lh->num_alloc_nodes;

	for (np = *n1; np != NULL; ) {
#ifndef OPENSSL_NO_HASH_COMP
		hash = np->hash;
#else
		hash = lh->hash(np->data);
		lh->num_hash_calls++;
#endif
		if ((hash % nni) != p) { /* move it */
			*n1 = (*n1)->next;
			np->next= *n2;
			*n2 = np;
		} else
			n1 = &((*n1)->next);
		np= *n1;
	}

	if ((lh->p) >= lh->pmax) {
		j = (int)lh->num_alloc_nodes * 2;
		n = reallocarray(lh->b, j, sizeof(LHASH_NODE *));
		if (n == NULL) {
/*			fputs("realloc error in lhash", stderr); */
			lh->error++;
			lh->p = 0;
			return;
		}
		/* else */
		for (i = (int)lh->num_alloc_nodes; i < j; i++)/* 26/02/92 eay */
			n[i] = NULL;			  /* 02/03/92 eay */
		lh->pmax = lh->num_alloc_nodes;
		lh->num_alloc_nodes = j;
		lh->num_expand_reallocs++;
		lh->p = 0;
		lh->b = n;
	}
}

static void
contract(_LHASH *lh)
{
	LHASH_NODE **n, *n1, *np;

	np = lh->b[lh->p + lh->pmax - 1];
	lh->b[lh->p+lh->pmax - 1] = NULL; /* 24/07-92 - eay - weird but :-( */
	if (lh->p == 0) {
		n = reallocarray(lh->b, lh->pmax, sizeof(LHASH_NODE *));
		if (n == NULL) {
/*			fputs("realloc error in lhash", stderr); */
			lh->error++;
			return;
		}
		lh->num_contract_reallocs++;
		lh->num_alloc_nodes /= 2;
		lh->pmax /= 2;
		lh->p = lh->pmax - 1;
		lh->b = n;
	} else
		lh->p--;

	lh->num_nodes--;
	lh->num_contracts++;

	n1 = lh->b[(int)lh->p];
	if (n1 == NULL)
		lh->b[(int)lh->p] = np;
	else {
		while (n1->next != NULL)
			n1 = n1->next;
		n1->next = np;
	}
}

static LHASH_NODE **getrn(_LHASH *lh, const void *data, unsigned long *rhash)
{
	LHASH_NODE **ret, *n1;
	unsigned long hash, nn;
	LHASH_COMP_FN_TYPE cf;

	hash = (*(lh->hash))(data);
	lh->num_hash_calls++;
	*rhash = hash;

	nn = hash % lh->pmax;
	if (nn < lh->p)
		nn = hash % lh->num_alloc_nodes;

	cf = lh->comp;
	ret = &(lh->b[(int)nn]);
	for (n1 = *ret; n1 != NULL; n1 = n1->next) {
#ifndef OPENSSL_NO_HASH_COMP
		lh->num_hash_comps++;
		if (n1->hash != hash) {
			ret = &(n1->next);
			continue;
		}
#endif
		lh->num_comp_calls++;
		if (cf(n1->data, data) == 0)
			break;
		ret = &(n1->next);
	}
	return (ret);
}

/* The following hash seems to work very well on normal text strings
 * no collisions on /usr/dict/words and it distributes on %2^n quite
 * well, not as good as MD5, but still good.
 */
unsigned long
lh_strhash(const char *c)
{
	unsigned long ret = 0;
	long n;
	unsigned long v;
	int r;

	if ((c == NULL) || (*c == '\0'))
		return (ret);
/*
	unsigned char b[16];
	MD5(c,strlen(c),b);
	return(b[0]|(b[1]<<8)|(b[2]<<16)|(b[3]<<24));
*/

	n = 0x100;
	while (*c) {
		v = n | (*c);
		n += 0x100;
		r = (int)((v >> 2) ^ v) & 0x0f;
		ret = (ret << r)|(ret >> (32 - r));
		ret &= 0xFFFFFFFFL;
		ret ^= v * v;
		c++;
	}
	return ((ret >> 16) ^ ret);
}

unsigned long
lh_num_items(const _LHASH *lh)
{
	return lh ? lh->num_items : 0;
}
